src/transformers/models/pop2piano/feature_extraction_pop2piano.py

```diff
@@ -0,0 +1,390 @@
+
+        return bpm, beat_times, confidence, estimates, essentia_beat_intervals
+
+    def interpolate_beat_times(self, beat_times, steps_per_beat, extend=False):
+        beat_times_function = scipy.interpolate.interp1d(
+            np.arange(beat_times.size),
+            beat_times,
+            bounds_error=False,
+            fill_value="extrapolate",
+        )
+        if extend:
+            beat_steps_8th = beat_times_function(np.linspace(0, beat_times.size, beat_times.size * steps_per_beat + 1))
+        else:
+            beat_steps_8th = beat_times_function(
+                np.linspace(0, beat_times.size - 1, beat_times.size * steps_per_beat - 1)
+            )
+        return beat_steps_8th
+
+    def extrapolate_beat_times(self, beat_times, n_extend=1):
+        beat_times_function = scipy.interpolate.interp1d(
+            np.arange(beat_times.size),
+            beat_times,
+            bounds_error=False,
+            fill_value="extrapolate",
+        )
+        ext_beats = beat_times_function(np.linspace(0, beat_times.size + n_extend - 1, beat_times.size + n_extend))
+
+        return ext_beats
+
+    def preprocess_mel(
+        self,
+        audio,
+        beatstep,
+        num_bars,
+        padding_value,
+    ):
+        """Preprocessing for log-mel spectrogram"""
+
+        num_steps = num_bars * 4
+        num_target_steps = len(beatstep)
+        extrapolated_beatstep = self.extrapolate_beat_times(beatstep, (num_bars + 1) * 4 + 1)
+
+        batch = []
+        for i in range(0, num_target_steps, num_steps):
+            start_idx = i
+            end_idx = min(i + num_steps, num_target_steps)
+
+            start_sample = int(extrapolated_beatstep[start_idx] * self.sampling_rate)
+            end_sample = int(extrapolated_beatstep[end_idx] * self.sampling_rate)
+            feature = audio[start_sample:end_sample]
+            batch.append(feature)
+        batch = pad_sequence(batch, batch_first=True, padding_value=padding_value)
+
+        return batch, extrapolated_beatstep
+
+    def single_preprocess(
+        self,
+        beatstep,
+        audio=None,
+    ):
+        """Preprocessing method for a single sequence."""
+
+        if audio is not None:
+            if len(audio.shape) != 1:
+                raise ValueError(
+                    f"Expected `audio` to be a single channel audio input of shape (n, ) but found shape {audio.shape}."
+                )
+
+        if beatstep[0] > 0.01:
+            logger.warning(f"`beatstep[0]` is not 0. All beatstep will be shifted by {beatstep[0]}.")
+            beatstep = beatstep - beatstep[0]
+
+        batch, extrapolated_beatstep = self.preprocess_mel(
+            audio,
+            beatstep,
+            num_bars=self.num_bars,
+            padding_value=self.padding_value,
+        )
+
+        return batch, extrapolated_beatstep
+
+    def pad(self, inputs: BatchFeature):
```

Comment:
Would it be simpler if we first pad the raw audio inputs to the same length, and **then** compute the log mel spectrograms on the padded audios? Currently, we compute the log mel spectrograms and **then** pad them. If we do it the first way, we can use the generic `self.pad` method and just forward the `attention_mask` we get onto the model?

Reply:
Even if we pad the audio files beforehand(to make them say of 300 secs each), the `extract_rhythm` will give different shapes of outputs even for each example! So we still need to pad them after that function.